content/blog/2019-09-25-c-implementation.md

```diff
@@ -0,0 +1,189 @@
+`src/interp.c`, first loops through the parsed bril program and notes the
+indices of labels. This was the simplest way to implement jumps and branches,
+which become a simple setting of the instruction pointer to the label's index
+(or an error if the label is not found in the label->index map). Actual
+instructions are implemented with a set of functions, one for each op code.
+Once the `op` is known, teh interpreter calls one of these functions., which
+fetches arguments, does the required manipulation, and stores its result in the
+right place (or in the case of effect operations, has teh correct effect). 
+
+## Op Code Implementation
+All instructions but `jmp`, `br`, `print`, and `const` have a trivial
+implementation. A function called `get_or_quit` fetches arguments from the
+source program and stores their values in a global array, sized to the max of
+the argument count of all non-print instructions (2). If the variable is not
+found in storage, an error is reported (containing the specific issue,
+incorrect variable, and instruction pointer value) and the program quits.
+Otherwise, the op implementing function calls `put` with the argument that
+implements the specific op code, which stores the result in the destination
+variable.. For example, the `add` op code is implemented
+as follows:
+
+```c
+static void op_add() {
+  get_or_quit();
+  put(mem_args[0]+mem_args[1]);
+}
+```
+Above, `mem_args` is the name of the global argument array. Thus, most op codes
+have 2 line implementations. Even `br`, the most complex op code, requires only
+16 lines to implement. 
+
+## Memory Implentation (and More, for Free)
+I needed a hash table implementation to implement variable storage, so I built
+it myself. It can be found in `src/table.{c,h}`. The dictionary uses open
+addressing and linear chaining. The hash function copies the hash used by
+java.lang.String’s hashCode() method. The `_hash` function computes
+a polynomial whos coeffecients are the integer values of the string's
+characters, evaluated at `x=31`. The polynomial is evaluated with Horner's
+method. The `table` maps string keys to `int64_t` values.  I represent bril
+integers and bril booleans with `int64_t` to avoid storing the type of bril
+variables. Since bril's typesystem only allows for those two types, this design
+is sufficient for now. Incorporating more types will be simple: I can extend
+the `table_elem` struct with a type bitfield/enum value.  For now, this simple
+table suffices. 
+
+The table code is also used for the label->index map and to store the mapping
+between string op codes and the index for that op code's implementing function
+in an array of function pointers.
+
+## Evaluation
+### Correctness
+I ran `cril` against the programs in `bril/test` and made sure the outputs
+matched `<program>.out`. My interpreter gave correct results on a fibonacci
+program posted in the class slack channel, on which the reference interpreter
```

Comment:
Remember that your audience is the world—and the entire world sadly isn't on Slack. So please include whatever's relevant about that code that your reader otherwise won't be able to see.